### PR TITLE
 [FLINK-33798][statebackend/rocksdb] automatically clean up rocksdb logs when the task exited.

### PR DESCRIPTION
**Pull request by liming30**
## What is the purpose of the change

 [FLINK-33798][statebackend/rocksdb] automatically clean up rocksdb logs when the task exited.

## Brief change log

When the task exits, delete all rocksdb log files with the same prefix in the relocated directory.

## Verifying this change

This change added tests and can be verified as follows:

- org.apache.flink.contrib.streaming.state.RocksDBStateBackendConfigTest#testCleanRelocatedDbLogs

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by liming30** [comment]
@flinkbot run azure

File: flink-state-backends/flink-statebackend-rocksdb/src/test/java/org/apache/flink/contrib/streaming/state/RocksDBStateBackendConfigTest.java [hunk 1]
```diff
@@ -359,6 +361,50 @@ public void testDbPathRelativePaths() throws Exception {
...
+
+        String relocatedDbLogPrefix =
+                RocksDBResourceContainer.resolveRelocatedDbLogPrefix(
+                        instanceRocksDBPath.getAbsolutePath());
+        java.nio.file.Path[] relocatedDbLogs;
+        try {
+            relocatedDbLogs = FileUtils.listDirectory(relocatedDBLogDir.toPath());
+            assertTrue(relocatedDbLogs[0].getFileName().startsWith(relocatedDbLogPrefix));
+            // add a rolled log file
+            Files.createTempFile(relocatedDBLogDir.toPath(), relocatedDbLogPrefix, ".suffix");
```
Comment (1996fanrui) [inline flink-state-backends/flink-statebackend-rocksdb/src/test/java/org/apache/flink/contrib/streaming/state/RocksDBStateBackendConfigTest.java:397]: Could we let rocksdb create LOG file and calling the close to clean it?

If so, we don't need to call the `resolveRelocatedDbLogPrefix`. Current test just tests create an file based on `resolveRelocatedDbLogPrefix`, and clean it based on `resolveRelocatedDbLogPrefix`.

If `resolveRelocatedDbLogPrefix` has some bugs or its rule isn't same with rocksdb side. This test cannot cover it.

Also, if we let rocksdb create LOG file, current test will fail if rocksdb log file rule is changed in the future.

WDYT?

File: flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBResourceContainer.java [hunk 2]
```diff
@@ -438,4 +451,44 @@ private File resolveFileLocation(String logFilePath) {
...
+    }
+
+    /**
+     * Resolve the prefix of rocksdb's log file name according to rocksdb's log file name rules. See
+     * https://github.com/ververica/frocksdb/blob/FRocksDB-6.20.3/file/filename.cc#L30.
+     *
+     * @param instanceRocksDBAbsolutePath The path where the rocksdb directory is located.
+     * @return Resolved rocksdb log name prefix.
+     */
+    public static String resolveRelocatedDbLogPrefix(String instanceRocksDBAbsolutePath) {
```
Comment (1996fanrui) [inline flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBResourceContainer.java:486]: Based on the last comment, I prefer this method is a private method.

Also, if it's public method in the end, please add `@VisibleForTesting` annotation.

**Review by 1996fanrui — COMMENTED** [review]
Thanks @liming30 for the fix!

Overall LGTM, I left 2 minor comments about testing, please take a look in your free time, thanks~

File: flink-state-backends/flink-statebackend-rocksdb/src/test/java/org/apache/flink/contrib/streaming/state/RocksDBStateBackendConfigTest.java (hunk 1 above)
Comment (liming30) [reply flink-state-backends/flink-statebackend-rocksdb/src/test/java/org/apache/flink/contrib/streaming/state/RocksDBStateBackendConfigTest.java]: Resolved.

File: flink-state-backends/flink-statebackend-rocksdb/src/test/java/org/apache/flink/contrib/streaming/state/RocksDBStateBackendConfigTest.java [hunk 3]
```diff
@@ -359,6 +365,58 @@ public void testDbPathRelativePaths() throws Exception {
...
+        final EmbeddedRocksDBStateBackend rocksDbBackend =
+                new EmbeddedRocksDBStateBackend().configure(conf, getClass().getClassLoader());
+        final String dbStoragePath = new Path(folder.toURI().toString()).toString();
+        rocksDbBackend.setDbStoragePath(dbStoragePath);
+
+        final MockEnvironment env = getMockEnvironment(tempFolder.newFolder());
+        RocksDBKeyedStateBackend<Integer> keyedBackend =
+                createKeyedStateBackend(rocksDbBackend, env, IntSerializer.INSTANCE);
+        // clear unused file
+        FileUtils.deleteFileOrDirectory(logFile);
```
Comment (1996fanrui) [inline flink-state-backends/flink-statebackend-rocksdb/src/test/java/org/apache/flink/contrib/streaming/state/RocksDBStateBackendConfigTest.java:390]: We don't need to clean the `taskManager.log`, right?

If so, we can check that only rocksdb log is cleaned.  (We should ensure rocksdb doesn't clean other logs.)

File: flink-state-backends/flink-statebackend-rocksdb/src/test/java/org/apache/flink/contrib/streaming/state/RocksDBStateBackendConfigTest.java [hunk 4]
```diff
@@ -359,6 +365,58 @@ public void testDbPathRelativePaths() throws Exception {
...
+
+        File instanceBasePath = keyedBackend.getInstanceBasePath();
+        File instanceRocksDBPath =
+                RocksDBKeyedStateBackendBuilder.getInstanceRocksDBPath(instanceBasePath);
+
+        // avoid tests without relocate.
+        Assume.assumeTrue(instanceRocksDBPath.getAbsolutePath().length() <= 255 - "_LOG".length());
```
Comment (1996fanrui) [inline flink-state-backends/flink-statebackend-rocksdb/src/test/java/org/apache/flink/contrib/streaming/state/RocksDBStateBackendConfigTest.java:397]: Could we ensure the relocation always take effect in current test?

If we cannot ensure it, this test will be ignored, and we can't  found this bug if the log clean logic is broken in the future.

WDYT?

Comment (liming30) [reply flink-state-backends/flink-statebackend-rocksdb/src/test/java/org/apache/flink/contrib/streaming/state/RocksDBStateBackendConfigTest.java:397]: Sorry, I cannot guarantee that the length of `instanceRocksDBPath` will be less than `255 - "_LOG".length()` in this test, which means that relocation may not occur. 

Forcibly specifying an absolute path may cause problems such as the path not existing and no permissions on the test machine. Is there any other way to ensure that relocation will definitely happen?

File: flink-state-backends/flink-statebackend-rocksdb/src/test/java/org/apache/flink/contrib/streaming/state/RocksDBStateBackendConfigTest.java (hunk 3 above)
Comment (liming30) [reply flink-state-backends/flink-statebackend-rocksdb/src/test/java/org/apache/flink/contrib/streaming/state/RocksDBStateBackendConfigTest.java:390]: Good idea, I will resolve it in the next commit.

File: flink-state-backends/flink-statebackend-rocksdb/src/test/java/org/apache/flink/contrib/streaming/state/RocksDBStateBackendConfigTest.java (hunk 4 above)
Comment (liming30) [reply flink-state-backends/flink-statebackend-rocksdb/src/test/java/org/apache/flink/contrib/streaming/state/RocksDBStateBackendConfigTest.java:397]: I noticed that the length limit issue was being addressed in `frocksdb`, which would ensure relocations occurred, but there was no further progress on the related PR. https://github.com/ververica/frocksdb/pull/66

**Review by 1996fanrui — APPROVED** [review]
Thanks @liming30 for the fix!

LGTM, merging~

**Pull request merged**